### ISSUE TITLE
Medaka consensus never finishes running

### ISSUE THREAD

**Issue by brijon** [nanoporetech/medaka#138]
Hi
I seem to have the exact same issue as described here: https://github.com/nanoporetech/medaka/issues/112
I was about to reply to the previous thread but it has recently been closed.
medaka consensus never finishes running and goes to sleep, I can still see the processes in htop but they are using 0% of each CPU they were previously running on.

My output looks like this and never continues (as in the previous persons post).
```
[02:20:05 - PWorker] 81.2% Done (0.0/0.0 Mbases) in 20.0s
[02:20:05 - PWorker] All done, 0 remainder regions.
[02:20:05 - DLoader] Initializing data loader
[02:20:05 - Sampler] Initializing sampler for consensus of region tig00002457:0-2427.
[02:20:05 - PWorker] Running inference for 0.0M draft bases.
[02:20:05 - Feature] Processed tig00002457:0.0-2426.0 (median depth 15.0)
[02:20:05 - Sampler] Took 0.31s to make features.
[02:20:15 - PWorker] All done, 0 remainder regions.
[02:20:15 - DLoader] Initializing data loader
[02:20:15 - Sampler] Initializing sampler for consensus of region tig00002470:0-3448.
[02:20:15 - PWorker] Running inference for 0.0M draft bases.
[02:20:15 - Feature] Processed tig00002470:0.0-3447.0 (median depth 13.0)
[02:20:15 - Sampler] Took 0.32s to make features.
[02:20:27 - PWorker] 77.3% Done (0.0/0.0 Mbases) in 12.0s
[02:20:27 - PWorker] All done, 0 remainder regions.
[02:20:27 - DLoader] Initializing data loader
[02:20:27 - PWorker] Running inference for 0.0M draft bases.
[02:20:27 - Sampler] Initializing sampler for consensus of region tig00002464:0-5165.
[02:20:27 - Feature] Processed tig00002464:0.0-5164.0 (median depth 21.0)
[02:20:27 - Sampler] Took 0.32s to make features.
[02:20:49 - PWorker] 86.6% Done (0.0/0.0 Mbases) in 22.4s
[02:20:49 - PWorker] All done, 0 remainder regions.
[02:20:49 - DLoader] Initializing data loader
[02:20:49 - PWorker] Running inference for 0.0M draft bases.
[02:20:49 - Sampler] Initializing sampler for consensus of region tig00002465:0-6080.
[02:20:50 - Feature] Processed tig00002465:0.0-6079.0 (median depth 11.0)
[02:20:50 - Sampler] Took 0.31s to make features.

```

I installed medaka 0.11.5 with conda , No GPU and have tried running it inside and outside of the virtualenv and consistently have the same problem. I have also tried running the medaka consensus script by itself with the same result.

My bam file is 17.9 gb.
Could this be a problem with the way I have medaka setup rather than a bug? Any insights much appreciated!
Best,
Briony

**Comment by mwykes** [comment]
Hi @brijon, as you mention, this is something other people have reported, however we have never been able to reproduce it internally, so don't have a means to debug it. 

We do know that the issue seems to be related to the chunking of consensus calculation as it only seems to happen for contigs which are shorter than the default consensus chunk size of 10kb. 

Could you try a shorter chunk size, such that all your contigs are longer than the chunk size? It looks like your contigs are all longer than 2kb, so that might be a good size to try. As the chunk size option is not exposed in `medaka_consensus`, you'll have to run 

`medaka consensus calls_to_draft.bam consensus_probs.hdf --chunk_len 2000`

You can then decode the probabilities to fasta using:

`medaka stitch consensus_probs.hdf consensus.fasta`

**Comment by brijon** [comment]
Thanks for the response @mwykes I will give this a go!

**Comment by fenderglass** [comment]
Hi,

I can confirm the same issue is often occurring when I am polishing a metagenomic assemblies (which have contigs as short as 500 bp). My workaround is to kill the medaka process and restart the same command line, which then continues from stitching and finishes successfully.

I have tried to adjust `--chunk-len` to 500 and ran:
`medaka consensus flye_hac/medaka/calls_to_draft.bam flye_hac/medaka/consensus_probs.hdf --batch_size 100 --threads 30  --chunk_len 500`. However, the size of `consensus_probs.hdf` file reduced 70-fold, as compared to the hdf I got originally - so something definitely went wrong.

Will be happy to help with testing on my side, if needed.

Best,
Mikhail

**Comment by cjw85** [comment]
@fenderglass I've reached out by email.

**Comment by cjw85** [comment]
A little bit of an update here.

We were again not able to reproduce this error, even with the dataset from @fenderglass; for most users the issue is non-deterministic.

One suspicion is that something to do with how medaka interacts with HDF files directly and via Tensorflow is to blame. The latest v1.1.2 medaka release upgrades both Tensorflow and the version of the h5py library used by python to interface with HDF files. In a subsequent release the RNN models will not be stored in HDF but Tensorflow's alternative format. If this suspicion is correct the problem may be allievated.

**Comment by MaestSi** [comment]
Hi, I am facing a similar issue with `medaka_consensus` program of Medaka v1.2.1.
The strange thing is that the same dataset completes without any errors on a server (with Ubuntu 20.04.1 LTS) while it always stops on another server (with Ubuntu 14.04.4 LTS).  Could the OS be responsbile for this issue, or is it just a coincidence?
Thanks,
Simone

**Comment by cjw85** [comment]
Collating some observations from other issues, setting the environment variable `HDF5_USE_FILE_LOCKING=FALSE` may help this scenario.

**Comment by MaestSi** [comment]
Hi @cjw85 , thanks for the information. Unfortunately, in my case Medaka v1.2.2 still doesn't work also after setting the env variable. However, I have switched back to version 1.0.3 which still works without any issues.
Best,
Simone

**Comment by cjw85** [comment]
@MaestSi,

Is it the case that v1.0.3 reliably works for you and v1.2.2 does not? You mentioned Ubuntu 14.04 above, which is at this point very nearly 7 years old. What happens for you on Ubuntu 20.04?

**Comment by MaestSi** [comment]
Hi @cjw85 ,
exactly, with Ubuntu 20.04 both versions work. Therefore, on one machine I kept Medaka v1.0.3, and on the newer one I updated to Medaka v1.2.2.
Simone

**Comment by cjw85** [comment]
I'm going to close this issue for now, setting of the `HDF5_USE_FILE_LOCKING` environment variable is now performed explicitely within medaka at runtime. We're not aware of any users for whom this hasn't resolved the issue; excepting @MaestSi's Ubuntu 14.04 case, which was end of life two years ago.
